Add YARD extensions for public API

marcotc opened the pull request:
This PR, extracted from #1778, provides the foundation to a richer YARD documentation that will support our public API documentation. This prepares us to tag public methods with a custom `@public_api` YARD tag.

This PR adds:
1. Generate documentation for our configuration DSL, mostly defined in [dd-trace-rb/lib/ddtrace/configuration/settings.rb](https://github.com/DataDog/dd-trace-rb/blob/40b107252be3fff18e51326b85f88853a02d5efc/lib/ddtrace/configuration/settings.rb#L45-L50). The `settings :name` and `option :name` will not generate classes and module that can be documented (classes are namespaced in a new `DSL` module, in order to avoid path conflicts):
1.1  <img width="1005" alt="Screen Shot 2021-12-08 at 3 58 10 PM" src="https://user-images.githubusercontent.com/583503/145310476-13ec6180-1c22-4c82-8841-440ade57c3bc.png">
1.2  The `c.analytics.enabled` option, for example, looks like this after being documented: <img width="1066" alt="Screen Shot 2021-12-08 at 4 06 27 PM" src="https://user-images.githubusercontent.com/583503/145310784-5340d75d-30c1-47f9-8c6a-4386f1e450c0.png">


2. Added a Public API list, to the top-left menu list. This list collects all modules, classes, and methods that are part of the public API and allows users to search for them dynamically in our docs:
<img width="314" alt="Screen Shot 2021-12-08 at 4 00 54 PM" src="https://user-images.githubusercontent.com/583503/145310980-9db8b7e1-7479-4270-8718-95491566a159.png">


3. Added a few extra documentation files, to allow us to directly link YARD docs to a more expanded explanation of a topic. The most useful case so far is to hyperlink to an explanation of the "Public API", which will be added to every `@public_api` tagged object. 
<img width="262" alt="Screen Shot 2021-12-08 at 3 58 58 PM" src="https://user-images.githubusercontent.com/583503/145311084-dea10d77-79b7-4cef-8348-a9745508510f.png">
You can see an example here:
<img width="360" alt="Screen Shot 2021-12-08 at 4 00 27 PM" src="https://user-images.githubusercontent.com/583503/145311316-b99bf28c-2716-4e35-87a1-6b1eb7ff60dd.png">

4. Cleaned up YARD and Rubocop Rake tasks, moving most options to their respected dot files. Using dot files as the main source of configuration allows external tools to correctly handle the `ddtrace`, without depending exclusively on our Rake tasks. One example is `yard server --reload`, which hosts a hot-reload YARD server that reflects documentation changes very quickly, but depends on having access to all options, which are now all in `.yardopts`.

ivoanjo reviewed (commented):
Ok, so the way I see it, the primary target of this documentation would be dd-trace-rb users, not us as the library developers (right?).

Thus, I think marking things as public api is a great start, but I think we should go further: we should outright hide (at least by default) everything that is not the public API.

Our library is open-source, so very advanced customers can always peek at the code or perhaps generate the docs with the private things included.

It appears to me that a default where we show everything and then individually mark the things that are available for use is really confusing, and it's not something I see usually in API docs. ([RSpec seems to show private some stuff, but marks it as such, rather than the reverse.](https://rspec.info/documentation/3.10/rspec-core/RSpec/Core/Configuration.html#profile_examples-instance_method))

Thus, I think our documentation should be built from the customer point of view: it's our way of saying "here's what we provide you so you can make the best out of Datadog", and guiding customers through that journey; right now what we have is more of a "here's all our guts exposed, see if you can figure out what you want out of it" 😉 

---

Separate, but related to the above, is while I think the DSL extension docs are impressive, I also don't think it's the right way either.

Is it just me or is the list on the right much clearer than the one in the docs?

![Screen Shot 2021-12-09 at 11 02 04](https://user-images.githubusercontent.com/2785847/145384663-59ef168e-0134-464d-827f-4c574bb220f2.png)

With my user hat on, and especially with a DSL, when I'm using it, I'm not really thinking of the settings as "there's a bunch of classes here I'm manipulating", I'm thinking "there's some grouping, and nested properties inside this grouping -- I'm effectively building a tree of hashes".

So while I wouldn't state it as a blocker, I don't think exposing the configuration in this way is very helpful. I would suggest looking into exposing it as a single flat file in the docs (I offer to help if that turns out to be thorny!).

marcotc commented:
> Ok, so the way I see it, the primary target of this documentation would be dd-trace-rb users, not us as the library developers (right?).

I think think we can't separate it today. I've never seen it separate before (two distinct but overlapping doc pages, one for people developing, another for users), and there's always the "canonical" https://www.rubydoc.info/gems/ddtrace page which we can only have one way there.

I don't think having two separate ones is really achievable today.

> Thus, I think marking things as public api is a great start, but I think we should go further: we should outright hide (at least by default) everything that is not the public API.

But then we hide everything, even from ourselves and library devs.

> It appears to me that a default where we show everything and then individually mark the things that are available for use is really confusing, and it's not something I see usually in API docs. ([RSpec seems to show private some stuff, but marks it as such, rather than the reverse.](https://rspec.info/documentation/3.10/rspec-core/RSpec/Core/Configuration.html#profile_examples-instance_method))

We could. We would still only tag @public_api methods/classes, as there is 10x more private methods than public, so it would be best for devs to mark public stuff only.
 
> Thus, I think our documentation should be built from the customer point of view: it's our way of saying "here's what we provide you so you can make the best out of Datadog", and guiding customers through that journey; right now what we have is more of a "here's all our guts exposed, see if you can figure out what you want out of it" 😉

There's a public API list of only methods that are part of the public API.
We can expose it differently, I don't know how to do it better given it's a shared doc today.

> Separate, but related to the above, is while I think the DSL extension docs are impressive, I also don't think it's the right way either.
> 
> Is it just me or is the list on the right much clearer than the one in the docs?
> 
> ![Screen Shot 2021-12-09 at 11 02 04](https://user-images.githubusercontent.com/2785847/145384663-59ef168e-0134-464d-827f-4c574bb220f2.png)
> 
> With my user hat on, and especially with a DSL, when I'm using it, I'm not really thinking of the settings as "there's a bunch of classes here I'm manipulating", I'm thinking "there's some grouping, and nested properties inside this grouping -- I'm effectively building a tree of hashes".
>
> So while I wouldn't state it as a blocker, I don't think exposing the configuration in this way is very helpful. I would suggest looking into exposing it as a single flat file in the docs (I offer to help if that turns out to be thorny!).

So, the configuration is a tricky one as `c.analytics.enabled` is an option, but in practical Ruby terms `#analytics` is a method in the Settings class, and `#enabled` is a method in the generated "analytics" object. I don't think hiding this is correct, given it's the truth.

I think having a "simple" document laying out all configuration options would be great, but I don't thinking hiding the DSL objects is correct.

marcotc commented on docs/AutoInstrumentation.md:
```diff
@@ -0,0 +1,30 @@
+## Additional configuration
+
+You can reconfigure, override, or disable any specific integration settings by adding
+a `Datadog.configure` call after `ddtrace/auto_instrument` is activated.
```
Done!

marcotc commented on Rakefile:
```diff
@@ -171,16 +171,19 @@ namespace :spec do
+  # as `.yardopts` can be read by external YARD tools, like the
+  # hot-reload YARD server `yard server --reload`.
+
   t.options += ['--title', "ddtrace #{Datadog::VERSION::STRING} documentation"]
```
Added a default title.

marcotc commented:
Ok, if we are to keep only one YARD doc page today, would you think it makes sense to just scrub all non-public API stuff?

I can flip it to that mode if that's what you are suggesting.

ivoanjo commented:
> Ok, if we are to keep only one YARD doc page today, would you think it makes sense to just scrub all non-public API stuff?
>
> I can flip it to that mode if that's what you are suggesting.

Yes, I think that's the way to go :)

> I think think we can't separate it today. I've never seen it separate before (two distinct but overlapping doc pages, one for people developing, another for users), and there's always the "canonical" https://www.rubydoc.info/gems/ddtrace page which we can only have one way there.
>
> I don't think having two separate ones is really achievable today.

I was thinking it could be as simple as an alternative rake task, `rake internal_api_docs` or something like that, that we could document in `docs/DevelopmentGuide.md`.

> We could. We would still only tag @public_api methods/classes, as there is 10x more private methods than public, so it would be best for devs to mark public stuff only.

I agree with you here; the default being private I think is a safer default, and forces us to think carefully every time we're about to expose something, rather than "oops I just added a class and it became part of the public API".

> So, the configuration is a tricky one as c.analytics.enabled is an option, but in practical Ruby terms #analytics is a method in the Settings class, and #enabled is a method in the generated "analytics" object. I don't think hiding this is correct, given it's the truth.
>
> I think having a "simple" document laying out all configuration options would be great, but I don't thinking hiding the DSL objects is correct.

I see your point. I guess I just think the current approach is not that useful and is more confusing than just referencing the usual docs or the source.

That said, since this is more about presentation of the docs, we can revisit it separately/later.

One note I would point out is that the yard extension seems to be a bit confused with the nesting -- I'm guessing having `Datadog::Settings::DSL::Profiling::DSL::Advanced::DSL::Endpoint::DSL::Collection` is not intended? ;)

marcotc commented:
> One note I would point out is that the yard extension seems to be a bit confused with the nesting -- I'm guessing having Datadog::Settings::DSL::Profiling::DSL::Advanced::DSL::Endpoint::DSL::Collection is not intended? ;)

It adds `DSL` to avoid namespace conflict.
Because it doesn't know if this is already a DSL, it adds it again.
I can remove the nested ones, if that's your concern.

marcotc commented:
@ivoanjo I've hidden all non-@public_api objects. I've also cleaned up the custom template/css for Public API, as those are not needed, given everything there is part of the public api now.

marcotc commented:
>  Because it doesn't know if this is already a DSL, it adds it again.
 I can remove the nested ones, if that's your concern.

Oh, I've done this one too.

marcotc commented:
@ivoanjo I'm going to merge this, as I addresses most of the points you mentioned. The reason for this is that I'm pushing to get all https://github.com/DataDog/dd-trace-rb/pull/1778 changes merged, and merging this one unblocks me to continue the work.

Please still feel free to let me know of any changes you see as essential, as I can address them as part of the work towards. https://github.com/DataDog/dd-trace-rb/pull/1778.

ivoanjo reviewed (commented):
Seems reasonable to merge this in, given that it's quite self-contained and we still are/will iterate on it.

One final note I'll leave here is that with the change to only display things marked with the public api tag the api docs are empty because the addition of the tag is coming in separately (e.g. #1811). 

This confused me for a bit until I figured that out :)
